### ISSUE TITLE
Wipeout rendering issues

### ISSUE THREAD

**Issue by aybe** [ata4/glrage#3]
Here's my preliminary take on rendering issues:

Files between PSX and ATI versions are binary identical for the ships ALLSH.PRM/CMP, it's quite puzzling as on why some of the ships are picking incorrect textures though they are correctly drawn, I guess that's unimportant for our problem.

SJR format are header-less RAW RGB555 textures, their width is either 256 or 512 pixels, height is probably inferred using width and file size, their width is hard-coded in the game for those I've seen.

Now for the missing polygons,
1. Their visibility seems tightly related to the refresh rate the game is currently at, going to windowed mode changes somewhat what's being drawn.
2. If you closely look at the starting lights in Altima VII, they are correctly rendered in their initial state (gray color). But as soon as they get tinted (red, orange, green) they do disappear. (shader issue ?)
3. There's a lot of jitter in regards to when a polygon will be drawn or not, stopping at the billboard before the long slope in Altima VII and looking at the right part of the billboards (made of 2 quads) will intermittently draw it. This game uses integers for its data and PSX did too (a rough guess) for 3D calculation, could there be some unintended Z-fighting / culled polygon being introduced while bringing the data from the game and transforming it for the OpenGL (being way more picky than a PSX) pipeline ? (could some fixed point arithmetic in the shader be of any help ?)
4. Artifacts definitely change when slowing down the whole thing (a poor man approach using Sleep in ContextImpl::renderBegin()) and you effectively see correct rendering (mostly). I am starting to wonder that this game might need a slow down (missing ATI logo, samples playing at wrong times) fix but I reckon this a tough/time consuming job.

Or a combination of multiple factors occurring at the same time ?

An interesting challenge (I guess) would be to try to get those quads presenting the text (in wireframe mode) to be really straight, as of now their drawing is missing many parts.

What about the other CIF games, do they exhibit the same sort of rendering issues ?

Sorry if it's a bit scattered through, I'll continue investigating and see if I can make any progress on the topic ... maybe that'll give you an idea or two and you might come up with a smarter approach :dart:

**Comment by aybe** [comment]
Here's some progress about tinting:

Start light colors are now correct, but that problem of missing polygons makes it very dim:

![2016-01-28 22_39_00-wipeout 95](https://cloud.githubusercontent.com/assets/1537591/12659614/0a31a8ea-c610-11e5-80e7-2f11418a7ca8.png)

![2016-01-28 22_42_25-wipeout 95](https://cloud.githubusercontent.com/assets/1537591/12659728/9506fa4c-c610-11e5-9b14-ebb9dad1373f.png)

Above screens shows https://github.com/ata4/glrage/blob/master/ati3dcif/shaders/ati3dcif.fsh#L86  replaced either by:
- `fragColor = texColor * vertColor;` 
- `fragColor = vertColor;`

Not sure if that might affect other games as I don't have them, for Wipeout AFAIK only lights and pads get modulated this way (B/W textures). Not sure why Arridos IV lights in tunnels entries' don't change color though...

**Comment by aybe** [comment]
There's another thing I've also encountered, for some reason we seem to be hooked in a color channel that serves as the color modulator for bonus pads, this happens for boosts as well as polygons above a billboard (see first encounter of either in track 1). This results in colors present in polygons that shouldn't be color-cycling.

Could there be some stuff incorrectly interpreted/unimplemented in the API ?

Also, I'm suspecting that those discard/distance uses in https://github.com/ata4/glrage/blob/master/ati3dcif/shaders/ati3dcif.fsh 'might' need further tweaking. 

I've also been trying to evaluate the fragment route taken by a pixel in CodeXL but that doesn't seem possible in OGL (http://stackoverflow.com/questions/2508818/how-to-debug-a-glsl-shader), unfortunately ... any ideas are welcome.

**Comment by ata4** [comment]
I fixed the texture lighting problems in 1762fb0, turns out I was using the wrong vertex order. ~~When flat shading is active, ATI3DCIF only uses the color of the first vertex to color a polygon while OpenGL only uses the last on default.~~ _Edit: both use the last vertex, so I just messed up the order in general._ In your edit, you used the interpolated vertex color (`vertColor`), which produced a black gradient across the triangulated quad, that's why it appeared so dim.

> Also, I'm suspecting that those discard/distance uses in https://github.com/ata4/glrage/blob/master/ati3dcif/shaders/ati3dcif.fsh 'might' need further tweaking. 

I don't think so. They've been extensively tested in Tomb Raider and Assault Rigs and work completely fine there. I've disabled fragment discarding in Wipeout as a test, it has nothing to do with the missing polygons. Even when rendering as GL_POINT, they're completely missing.

> What about the other CIF games, do they exhibit the same sort of rendering issues ?

Not any missing polygons, no. Assault Rigs has a minor jittering problem, since it also uses integer arithmetic, but it's not nearly as severe as in Wipeout. Tomb Raider used to have this problem as well, but since it adapts itself to the current resolution, it was automatically fixed once I figured out how to enable native HD resolutions.

> Artifacts definitely change when slowing down the whole thing (a poor man approach using Sleep in ContextImpl::renderBegin()) and you effectively see correct rendering (mostly). I am starting to wonder that this game might need a slow down (missing ATI logo, samples playing at wrong times) fix but I reckon this a tough/time consuming job.

I did a quick test with Sleep, but it doesn't appear to have any effect, as far as I can tell. At least half of the skybox and large parts of rocks in the environment of the first level still are missing even when playing with 10 FPS or less.

**Comment by ata4** [comment]
Okay, I figured out why many polygons are missing. Looks like some are rendered _behind_ the camera for some reason, so I had to update the projection matrix to include these, too. Wipeout now looks a **lot** better! There are still some minor clipping issues, but they appear on the original ATI Rage, too.

**Comment by aybe** [comment]
Amazing :+1: :+1: :+1:

**Comment by ata4** [comment]
I just uploaded r4, which also fixes some missing background images in the title screen and menu. I think Wipeout now runs mostly fine without serious glitches. Hope it's a help for your remake project. :smile:

**Comment by aybe** [comment]
Great !

I did figure these incorrect screens while playing the championship, tried to find the cause but no luck. Been trying to grab ideas from the Tomb Raider patches without success, it's quite amazing the amount of fixes you've put in for this game :+1:

And yes, the ATI version is quite helpful in understanding a few bits of the game, the high resolution definitely shows more details, i.e. I now exactly know what are the particles emitted by missiles, in the PSX things are a bit confusing due to the low resolution. Also the patches allows to very quickly get in the game in comparison to an emulator to try things out. I just resumed work on WXX-Rebirth now, I hope I'll be able to show significant advancement soon.

Thank you.
